tests/nodeos_read_terminate_at_block_test.py

```diff
@@ -0,0 +1,316 @@
+         4:"--read-mode irreversible",
+         6:"--read-mode irreversible",
+         9:"--plugin eosio::producer_api_plugin"})
+
+   producingNodeId = 0
+   producingNode = cluster.getNode(producingNodeId)
+
+   def stopProdNode():
+      if not producingNode.killed:
+         producingNode.kill(signal.SIGTERM)
+
+   def startProdNode():
+      if producingNode.killed:
+         relaunchNode(producingNode)
+
+   # Give some time for it to produce, so lib is advancing
+   waitForBlksProducedAndLibAdvanced()
+
+   # Kill all nodes, so we can test all node in isolated environment
+   for clusterNode in cluster.nodes:
+      clusterNode.kill(signal.SIGTERM)
+   cluster.biosNode.kill(signal.SIGTERM)
+
+   # Wrapper function to execute test
+   # This wrapper function will resurrect the node to be tested, and shut it down by the end of the test
+   def executeTest(nodeIdOfNodeToTest, runTestScenario):
+      testResult = False
+      resultDesc = None
+      try:
+         # Relaunch killed node so it can be used for the test
+         nodeToTest = cluster.getNode(nodeIdOfNodeToTest)
+         Utils.Print("Re-launch node #{} to excute test scenario: {}".format(nodeIdOfNodeToTest, runTestScenario.__name__))
+         relaunchNode(nodeToTest, relaunchAssertMessage="Fail to relaunch before running test scenario")
+
+         # Run test scenario
+         runTestScenario(nodeIdOfNodeToTest, nodeToTest)
+         resultDesc = "!!!TEST CASE #{} ({}) IS SUCCESSFUL".format(nodeIdOfNodeToTest, runTestScenario.__name__)
+         testResult = True
+      except Exception as e:
+         resultDesc = "!!!BUG IS CONFIRMED ON TEST CASE #{} ({}): {}".format(nodeIdOfNodeToTest, runTestScenario.__name__, e)
+      finally:
+         Utils.Print(resultDesc)
+         testResultMsgs.append(resultDesc)
+         # Kill node after use
+         if not nodeToTest.killed: nodeToTest.kill(signal.SIGTERM)
+      return testResult
+
+
+   # 1th test case: Replay in irreversible mode with reversible blks while connected to producing node
+   # Expectation: Node replays and launches successfully
+   #              and the head and lib should be no less than termAtBlockNum
+   def replayInIrrModeWithRevBlksAndConnectedToProdNodeTermAtBlock(nodeIdOfNodeToTest, nodeToTest):
+      try:
+         startProdNode()
+         # Kill node and replay in irreversible mode
+         nodeToTest.kill(signal.SIGTERM)
+         waitForBlksProducedAndLibAdvanced() # Wait
+         infoprod = producingNode.getInfo()
+         libprod = int(infoprod["last_irreversible_block_num"])
+         termAtBlockNum = libprod + termAtFutureBlockNum
+        
+         relaunchTestNode(nodeToTest, chainArg=" --read-mode irreversible  --terminate-at-block=" + str(termAtBlockNum))
+         Utils.Print("Test node begin recevie from producing node and is expected to stop at or little bigger than block number " + str(termAtBlockNum) )
+
+         while True :
+             info = nodeToTest.getInfo()
```

Comment:
I think you can just use here `Node.waitForHeadToAdvance`

Reply:
Also, if you need to know exactly the very last head & lib, you can parse blocks.log, there are methods for this:
`Cluster.getBlockLog`.
That way you'll eliminate any flakiness.

Reply:
There does seem to be the occasional failure when using getInfo. When parsing the block log how do you distinguish between head and lib?  I can see the last block_num corresponds to the output of get info, and seems match the head.  Also, is there a way to get the fork_db_head_block_num from the block log?  Is there a good reference for this? Thanks

Reply:
The blocklog only has lib in it. You can ask the controller for fork_db_head_block_num.

Reply:
Still learning, is the controller in this context a producer? I think there's a controller module as well? Thanks

Reply:
Sorry for being terse. `cleos get info` returns this information:

```
{
  "server_version": "eb71ad75",
  "chain_id": "3dddd240fb31aea58e7b19efdbf4c33d865d9f0bdfb11111b32468fbde4e35dd",
  "head_block_num": 72336,
  "last_irreversible_block_num": 72335,
  "last_irreversible_block_id": "00011a8f84e867fe04bd5a44c7f0de40c37a0e3bb781167fea43cde407e2b0cc",
  "head_block_id": "00011a9068c7cf12dbc4464d6bb9b0dd3b4daf9320eaceeb9ba1c61665f661fe",
  "head_block_time": "2021-08-20T19:57:19.000",
  "head_block_producer": "eosio",
  "virtual_block_cpu_limit": 450000000,
  "virtual_block_net_limit": "10485760000",
  "block_cpu_limit": 450000,
  "block_net_limit": 10485760,
  "server_version_string": "v2.1.0",
  "fork_db_head_block_num": 72336,
  "fork_db_head_block_id": "00011a9068c7cf12dbc4464d6bb9b0dd3b4daf9320eaceeb9ba1c61665f661fe",
  "server_full_version_string": "v2.1.0-eb71ad7560e3293c2ed152cca459789f6abf9564",
  "last_irreversible_block_time": "2021-08-20T19:57:18.500"
}
```

The fork db data is in there, it is retrieved from nodeos state. If python `getInfo()` fails then it should just be retried.

Reply:
I don't like race conditions here, executing getInfo until nodeos shutdown.. this test may become flaky.
what if we won't catch the very last block before nodeos shutdown?
`Cluster.getBlockLog` is using `eosio-blocklog' and later is using not just `blocks.log` but also `fork_db` that contains reversible blocks

Reply:
While the tests seem to pass fine, if I print the last element of block log vs. getInfo(), there's normally an additional block in the block log.

Assuming I drop fork_db_head_block number from the test to use getBlockLog(), I'm not clear how to retrieve the lib vs the head without using getInfo().

The log seems to be a list of json objects, and when comparing the log to the last successful getInfo() call, the log seems to include the head block (or maybe it just became the next lib and getInfo() missed it).

Reply:
that looks unfortunate limitation of eosio-blocklog. In that case we could skip checking LIB at all. I don't see any benefit of checking it here.